File: ios/PortalApiClient.swift
import Foundation

private func toJSONValue<T: Encodable>(_ value: T) -> JSONValue {
    do {
        let enc = JSONEncoder()
        let data = try enc.encode(value)
        let decoded = try JSONDecoder().decode(JSONValue.self, from: data)
        return decoded
    } catch {
        return .object(["_encodeError": .string(error.localizedDescription)])
    }
}

private func decodeAnyJsonValue(_ data: Data) -> JSONValue? {
    if data.isEmpty { return nil }
    if let j = try? JSONDecoder().decode(JSONValue.self, from: data) { return j }
    if let s = String(data: data, encoding: .utf8) { return .string(s) }
    return nil
}

private func getErrorMessage(from body: JSONValue?) -> String {
    guard let body else { return "Request failed" }
    switch body {
    case .string(let s):
        return s
    case .object(let o):
        if case .string(let msg)? = o["message"] { return msg }
        if case .string(let err)? = o["error"] { return err }
        if case .string(let details)? = o["details"] { return details }
        return "Request failed"
    default:
        return "Request failed"
    }
}

private func getErrorCode(from body: JSONValue?) -> String? {
    guard let body else { return nil }
    guard case .object(let o) = body else { return nil }
    guard case .string(let code)? = o["code"] else { return nil }
    return code
}

private func nowIso() -> String {
    ISO8601DateFormatter().string(from: Date())
}

private func requestId() -> String {
    UUID().uuidString.lowercased()
}

private func escapeSingleQuotes(_ s: String) -> String {
    // mimic bash-safe escaping used in many curl examples
    s.replacingOccurrences(of: "'", with: "'\\''")
}

private func buildCurl(url: String, hasAdmin: Bool, payloadJson: String) -> String {
    var headers = [
        "-H 'Content-Type: application/json'",
        "-H 'Authorization: Bearer <ID_TOKEN>'"
    ]
    if hasAdmin {
        headers.append("-H 'x-admin-token: <ADMIN_TOKEN>'")
    }
    return "curl -X POST \(headers.joined(separator: " ")) -d '\(escapeSingleQuotes(payloadJson))' '\(url)'"
}

// MARK: - Client errors

enum PortalApiClientError: Error, CustomStringConvertible {
    case invalidURL(String, meta: PortalApiMeta?)
    case encodingFailed(String, meta: PortalApiMeta?)
    case networkFailed(String, meta: PortalApiMeta?)
    case serverError(status: Int, message: String, envelope: PortalApiErrorEnvelope?, meta: PortalApiMeta)
    case decodingFailed(String, meta: PortalApiMeta?)

    var description: String {
        switch self {
        case .invalidURL(let s, _): return "Invalid URL: \(s)"
        case .encodingFailed(let s, _): return "Encoding failed: \(s)"
        case .networkFailed(let s, _): return "Network failed: \(s)"
        case .serverError(let status, let message, _, _): return "HTTP \(status): \(message)"
        case .decodingFailed(let s, _): return "Decoding failed: \(s)"
        }
    }
}

// MARK: - Portal API client

final class PortalApiClient {
    struct Config {
        let baseUrl: String
        init(baseUrl: String) {
            self.baseUrl = baseUrl.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        }
    }

    struct CallResult<T: Decodable> {
        let data: T
        let meta: PortalApiMeta
    }

    private let config: Config
    private let session: URLSession

    init(config: Config, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    // Generic post (mirrors portalApi.ts)
    func post<TReq: Encodable, TResp: Decodable>(
        fn: String,
        idToken: String,
        adminToken: String?,
        payload: TReq,
        respType: TResp.Type = TResp.self
    ) async throws -> CallResult<TResp> {

        let fullUrl = "\(config.baseUrl)/\(fn)"
        guard let url = URL(string: fullUrl) else {
            throw PortalApiClientError.invalidURL(fullUrl, meta: nil)
        }

        // Encode payload (pretty + stable keys helps debugging)
        let jsonData: Data
        let payloadJson: String
        do {
            let enc = JSONEncoder()
            enc.outputFormatting = [.prettyPrinted, .sortedKeys]
            jsonData = try enc.encode(payload)
            payloadJson = String(data: jsonData, encoding: .utf8) ?? "{}"
        } catch {
            throw PortalApiClientError.encodingFailed(error.localizedDescription, meta: nil)
        }

        var meta = PortalApiMeta(
            atIso: nowIso(),
            requestId: requestId(),
            fn: fn,
            url: fullUrl,
            payload: toJSONValue(payload),
            curlExample: buildCurl(url: fullUrl, hasAdmin: (adminToken?.isEmpty == false), payloadJson: payloadJson),
            status: nil,
            ok: nil,
            response: nil,
            error: nil,
            message: nil,
            code: nil
        )

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        if let adminToken, !adminToken.isEmpty {
            req.setValue(adminToken, forHTTPHeaderField: "x-admin-token")
        }
        req.httpBody = jsonData

        let data: Data
        let http: HTTPURLResponse
        do {
            let (d, r) = try await session.data(for: req)
            data = d
            guard let h = r as? HTTPURLResponse else {
                meta.ok = false
                meta.error = "Non-HTTP response"
                meta.message = "Non-HTTP response"
                throw PortalApiClientError.networkFailed("Non-HTTP response", meta: meta)
            }
            http = h
        } catch {
            let msg = error.localizedDescription
            meta.ok = false
            meta.error = msg
            meta.message = msg
            throw PortalApiClientError.networkFailed(msg, meta: meta)
        }

        meta.status = http.statusCode
        meta.ok = (200...299).contains(http.statusCode)
        meta.response = decodeAnyJsonValue(data)

        // Handle non-2xx as error envelope if possible
        if !(200...299).contains(http.statusCode) {
            let code = getErrorCode(from: meta.response)
            let msg = getErrorMessage(from: meta.response)
            let env = try? JSONDecoder().decode(PortalApiErrorEnvelope.self, from: data)
            meta.error = msg
            meta.message = msg
            meta.code = code ?? env?.code
            throw PortalApiClientError.serverError(status: http.statusCode, message: msg, envelope: env, meta: meta)
        }

        // Decode success
        do {
            let decoded = try JSONDecoder().decode(TResp.self, from: data)
            return CallResult(data: decoded, meta: meta)
        } catch {
            meta.ok = false
            meta.error = "Decoding failed"
            meta.message = error.localizedDescription
            throw PortalApiClientError.decodingFailed(error.localizedDescription, meta: meta)
        }
    }

    // MARK: - Typed helpers (optional convenience)

    func createBatch(idToken: String, adminToken: String?, payload: CreateBatchRequest) async throws -> CallResult<CreateBatchResponse> {
        try await post(fn: "createBatch", idToken: idToken, adminToken: adminToken, payload: payload, respType: CreateBatchResponse.self)
    }

    func pickedUpAndClose(idToken: String, adminToken: String?, payload: PickedUpAndCloseRequest) async throws -> CallResult<PickedUpAndCloseResponse> {
        try await post(fn: "pickedUpAndClose", idToken: idToken, adminToken: adminToken, payload: payload, respType: PickedUpAndCloseResponse.self)
    }

    func continueJourney(idToken: String, adminToken: String?, payload: ContinueJourneyRequest) async throws -> CallResult<ContinueJourneyResponse> {
        try await post(fn: "continueJourney", idToken: idToken, adminToken: adminToken, payload: payload, respType: ContinueJourneyResponse.self)
    }

    func listMaterialsProducts(idToken: String, adminToken: String?, payload: ListMaterialsProductsRequest) async throws -> CallResult<ListMaterialsProductsResponse> {
        try await post(fn: "listMaterialsProducts", idToken: idToken, adminToken: adminToken, payload: payload, respType: ListMaterialsProductsResponse.self)
    }

    func createMaterialsCheckoutSession(idToken: String, adminToken: String?, payload: CreateMaterialsCheckoutSessionRequest) async throws -> CallResult<CreateMaterialsCheckoutSessionResponse> {
        try await post(fn: "createMaterialsCheckoutSession", idToken: idToken, adminToken: adminToken, payload: payload, respType: CreateMaterialsCheckoutSessionResponse.self)
    }

    func seedMaterialsCatalog(idToken: String, adminToken: String?, payload: SeedMaterialsCatalogRequest) async throws -> CallResult<SeedMaterialsCatalogResponse> {
        try await post(fn: "seedMaterialsCatalog", idToken: idToken, adminToken: adminToken, payload: payload, respType: SeedMaterialsCatalogResponse.self)
    }

    func createReservation(idToken: String, adminToken: String?, payload: CreateReservationRequest) async throws -> CallResult<CreateReservationResponse> {
        try await post(fn: "createReservation", idToken: idToken, adminToken: adminToken, payload: payload, respType: CreateReservationResponse.self)
    }

    func listEvents(idToken: String, adminToken: String?, payload: ListEventsRequest) async throws -> CallResult<ListEventsResponse> {
        try await post(fn: "listEvents", idToken: idToken, adminToken: adminToken, payload: payload, respType: ListEventsResponse.self)
    }

    func listEventSignups(idToken: String, adminToken: String?, payload: ListEventSignupsRequest) async throws -> CallResult<ListEventSignupsResponse> {
        try await post(fn: "listEventSignups", idToken: idToken, adminToken: adminToken, payload: payload, respType: ListEventSignupsResponse.self)
    }

    func listBillingSummary(idToken: String, adminToken: String?, payload: ListBillingSummaryRequest) async throws -> CallResult<BillingSummaryResponse> {
        try await post(fn: "listBillingSummary", idToken: idToken, adminToken: adminToken, payload: payload, respType: BillingSummaryResponse.self)
    }

    func getEvent(idToken: String, adminToken: String?, payload: GetEventRequest) async throws -> CallResult<GetEventResponse> {
        try await post(fn: "getEvent", idToken: idToken, adminToken: adminToken, payload: payload, respType: GetEventResponse.self)
    }

    func createEvent(idToken: String, adminToken: String?, payload: CreateEventRequest) async throws -> CallResult<CreateEventResponse> {
        try await post(fn: "createEvent", idToken: idToken, adminToken: adminToken, payload: payload, respType: CreateEventResponse.self)
    }

    func publishEvent(idToken: String, adminToken: String?, payload: PublishEventRequest) async throws -> CallResult<PublishEventResponse> {
        try await post(fn: "publishEvent", idToken: idToken, adminToken: adminToken, payload: payload, respType: PublishEventResponse.self)
    }

    func signupForEvent(idToken: String, adminToken: String?, payload: SignupForEventRequest) async throws -> CallResult<SignupForEventResponse> {
        try await post(fn: "signupForEvent", idToken: idToken, adminToken: adminToken, payload: payload, respType: SignupForEventResponse.self)
    }

    func cancelEventSignup(idToken: String, adminToken: String?, payload: CancelEventSignupRequest) async throws -> CallResult<CancelEventSignupResponse> {
        try await post(fn: "cancelEventSignup", idToken: idToken, adminToken: adminToken, payload: payload, respType: CancelEventSignupResponse.self)
    }

    func claimEventOffer(idToken: String, adminToken: String?, payload: ClaimEventOfferRequest) async throws -> CallResult<ClaimEventOfferResponse> {
        try await post(fn: "claimEventOffer", idToken: idToken, adminToken: adminToken, payload: payload, respType: ClaimEventOfferResponse.self)
    }

    func checkInEvent(idToken: String, adminToken: String?, payload: CheckInEventRequest) async throws -> CallResult<CheckInEventResponse> {
        try await post(fn: "checkInEvent", idToken: idToken, adminToken: adminToken, payload: payload, respType: CheckInEventResponse.self)
    }

    func createEventCheckoutSession(idToken: String, adminToken: String?, payload: CreateEventCheckoutSessionRequest) async throws -> CallResult<CreateEventCheckoutSessionResponse> {
        try await post(fn: "createEventCheckoutSession", idToken: idToken, adminToken: adminToken, payload: payload, respType: CreateEventCheckoutSessionResponse.self)
    }

    func importLibraryIsbns(idToken: String, adminToken: String?, payload: ImportLibraryIsbnsRequest) async throws -> CallResult<ImportLibraryIsbnsResponse> {
        try await post(fn: "importLibraryIsbns", idToken: idToken, adminToken: adminToken, payload: payload, respType: ImportLibraryIsbnsResponse.self)
    }

    func registerDeviceToken(idToken: String, adminToken: String?, payload: RegisterDeviceTokenRequest) async throws -> CallResult<RegisterDeviceTokenResponse> {
        try await post(fn: "registerDeviceToken", idToken: idToken, adminToken: adminToken, payload: payload, respType: RegisterDeviceTokenResponse.self)
    }

    func unregisterDeviceToken(idToken: String, adminToken: String?, payload: UnregisterDeviceTokenRequest) async throws -> CallResult<UnregisterDeviceTokenResponse> {
        try await post(fn: "unregisterDeviceToken", idToken: idToken, adminToken: adminToken, payload: payload, respType: UnregisterDeviceTokenResponse.self)
    }

    func runNotificationFailureDrill(idToken: String, adminToken: String?, payload: RunNotificationFailureDrillRequest) async throws -> CallResult<RunNotificationFailureDrillResponse> {
        try await post(fn: "runNotificationFailureDrill", idToken: idToken, adminToken: adminToken, payload: payload, respType: RunNotificationFailureDrillResponse.self)
    }

    func runNotificationMetricsAggregationNow(idToken: String, adminToken: String?, payload: RunNotificationMetricsAggregationNowRequest = .init()) async throws -> CallResult<RunNotificationMetricsAggregationNowResponse> {
        try await post(fn: "runNotificationMetricsAggregationNow", idToken: idToken, adminToken: adminToken, payload: payload, respType: RunNotificationMetricsAggregationNowResponse.self)
    }
}

File: android/app/src/main/java/com/monsoonfire/portal/reference/PortalApiClient.kt
package com.monsoonfire.portal.reference

import java.io.IOException
import java.util.UUID
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive

class PortalApiException(message: String, val meta: PortalApiMeta) : Exception(message)

class PortalApiClient(
    private val config: Config,
    private val client: OkHttpClient = OkHttpClient(),
    private val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
        explicitNulls = false
    }
) {
    data class Config(val baseUrl: String)

    data class CallResult<T>(val data: T, val meta: PortalApiMeta)

    var lastMeta: PortalApiMeta? = null
        private set

    private fun nowIso(): String = java.time.Instant.now().toString()

    private fun requestId(): String = UUID.randomUUID().toString().lowercase()

    private fun escapeSingleQuotes(value: String): String = value.replace("'", "'\\''")

    private fun buildCurl(url: String, hasAdmin: Boolean, payloadJson: String): String {
        val headers = mutableListOf(
            "-H 'Content-Type: application/json'",
            "-H 'Authorization: Bearer <ID_TOKEN>'"
        )
        if (hasAdmin) headers.add("-H 'x-admin-token: <ADMIN_TOKEN>'")
        return "curl -X POST ${headers.joinToString(" ")} -d '${escapeSingleQuotes(payloadJson)}' '$url'"
    }

    private fun parseJsonElement(body: String): JsonElement? {
        if (body.isBlank()) return null
        return try {
            json.parseToJsonElement(body)
        } catch (_: Exception) {
            JsonPrimitive(body)
        }
    }

    @Throws(PortalApiException::class)
    fun <TReq, TResp> post(
        fn: String,
        idToken: String,
        adminToken: String?,
        payload: TReq,
        reqSerializer: KSerializer<TReq>,
        respSerializer: KSerializer<TResp>
    ): CallResult<TResp> {
        val base = config.baseUrl.trim().trimEnd('/')
        val url = "$base/$fn"

        val payloadJson = try {
            json.encodeToString(reqSerializer, payload)
        } catch (e: Exception) {
            val meta = PortalApiMeta(
                atIso = nowIso(),
                requestId = requestId(),
                fn = fn,
                url = url,
                payload = JsonPrimitive("<encode failed>"),
                curlExample = null,
                status = null,
                ok = false,
                response = null,
                error = "Encoding failed",
                message = e.message ?: "Encoding failed",
                code = null
            )
            lastMeta = meta
            throw PortalApiException(meta.message ?: "Encoding failed", meta)
        }

        val metaStart = PortalApiMeta(
            atIso = nowIso(),
            requestId = requestId(),
            fn = fn,
            url = url,
            payload = parseJsonElement(payloadJson) ?: JsonPrimitive(payloadJson),
            curlExample = buildCurl(url, !adminToken.isNullOrBlank(), payloadJson),
            status = null,
            ok = null,
            response = null,
            error = null,
            message = null,
            code = null
        )

        val requestBody = payloadJson.toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .header("Authorization", "Bearer $idToken")
            .apply {
                if (!adminToken.isNullOrBlank()) {
                    header("x-admin-token", adminToken)
                }
            }
            .build()

        try {
            client.newCall(request).execute().use { response ->
                return handleResponse(response, metaStart, respSerializer)
            }
        } catch (e: IOException) {
            val meta = metaStart.copy(
                ok = false,
                error = "Failed to fetch",
                message = e.message ?: "Failed to fetch"
            )
            lastMeta = meta
            throw PortalApiException(meta.message ?: "Failed to fetch", meta)
        }
    }

    private fun <TResp> handleResponse(
        response: Response,
        metaStart: PortalApiMeta,
        respSerializer: KSerializer<TResp>
    ): CallResult<TResp> {
        val bodyString = response.body?.string().orEmpty()
        val bodyJson = parseJsonElement(bodyString)

        val metaDone = metaStart.copy(
            status = response.code,
            ok = response.isSuccessful,
            response = bodyJson
        )

        if (!response.isSuccessful) {
            val env = try {
                json.decodeFromString(PortalApiErrorEnvelope.serializer(), bodyString)
            } catch (_: Exception) {
                null
            }
            val msg = env?.message ?: env?.error ?: "Request failed"
            val metaFail = metaDone.copy(
                error = msg,
                message = msg,
                code = env?.code
            )
            lastMeta = metaFail
            throw PortalApiException(msg, metaFail)
        }

        val decoded = try {
            json.decodeFromString(respSerializer, bodyString)
        } catch (e: Exception) {
            val metaFail = metaDone.copy(
                ok = false,
                error = "Decoding failed",
                message = e.message ?: "Decoding failed"
            )
            lastMeta = metaFail
            throw PortalApiException(metaFail.message ?: "Decoding failed", metaFail)
        }

        lastMeta = metaDone
        return CallResult(decoded, metaDone)
    }

    fun createBatch(
        idToken: String,
        adminToken: String?,
        payload: CreateBatchRequest
    ): CallResult<CreateBatchResponse> {
        return post(
            fn = "createBatch",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CreateBatchRequest.serializer(),
            respSerializer = CreateBatchResponse.serializer()
        )
    }

    fun pickedUpAndClose(
        idToken: String,
        adminToken: String?,
        payload: PickedUpAndCloseRequest
    ): CallResult<PickedUpAndCloseResponse> {
        return post(
            fn = "pickedUpAndClose",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = PickedUpAndCloseRequest.serializer(),
            respSerializer = PickedUpAndCloseResponse.serializer()
        )
    }

    fun continueJourney(
        idToken: String,
        adminToken: String?,
        payload: ContinueJourneyRequest
    ): CallResult<ContinueJourneyResponse> {
        return post(
            fn = "continueJourney",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ContinueJourneyRequest.serializer(),
            respSerializer = ContinueJourneyResponse.serializer()
        )
    }

    fun listMaterialsProducts(
        idToken: String,
        adminToken: String?,
        payload: ListMaterialsProductsRequest
    ): CallResult<ListMaterialsProductsResponse> {
        return post(
            fn = "listMaterialsProducts",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ListMaterialsProductsRequest.serializer(),
            respSerializer = ListMaterialsProductsResponse.serializer()
        )
    }

    fun createMaterialsCheckoutSession(
        idToken: String,
        adminToken: String?,
        payload: CreateMaterialsCheckoutSessionRequest
    ): CallResult<CreateMaterialsCheckoutSessionResponse> {
        return post(
            fn = "createMaterialsCheckoutSession",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CreateMaterialsCheckoutSessionRequest.serializer(),
            respSerializer = CreateMaterialsCheckoutSessionResponse.serializer()
        )
    }

    fun seedMaterialsCatalog(
        idToken: String,
        adminToken: String?,
        payload: SeedMaterialsCatalogRequest
    ): CallResult<SeedMaterialsCatalogResponse> {
        return post(
            fn = "seedMaterialsCatalog",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = SeedMaterialsCatalogRequest.serializer(),
            respSerializer = SeedMaterialsCatalogResponse.serializer()
        )
    }

    fun createReservation(
        idToken: String,
        adminToken: String?,
        payload: CreateReservationRequest
    ): CallResult<CreateReservationResponse> {
        return post(
            fn = "createReservation",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CreateReservationRequest.serializer(),
            respSerializer = CreateReservationResponse.serializer()
        )
    }

    fun listEvents(
        idToken: String,
        adminToken: String?,
        payload: ListEventsRequest
    ): CallResult<ListEventsResponse> {
        return post(
            fn = "listEvents",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ListEventsRequest.serializer(),
            respSerializer = ListEventsResponse.serializer()
        )
    }

    fun listEventSignups(
        idToken: String,
        adminToken: String?,
        payload: ListEventSignupsRequest
    ): CallResult<ListEventSignupsResponse> {
        return post(
            fn = "listEventSignups",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ListEventSignupsRequest.serializer(),
            respSerializer = ListEventSignupsResponse.serializer()
        )
    }

    fun listBillingSummary(
        idToken: String,
        adminToken: String?,
        payload: ListBillingSummaryRequest
    ): CallResult<BillingSummaryResponse> {
        return post(
            fn = "listBillingSummary",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ListBillingSummaryRequest.serializer(),
            respSerializer = BillingSummaryResponse.serializer()
        )
    }

    fun getEvent(
        idToken: String,
        adminToken: String?,
        payload: GetEventRequest
    ): CallResult<GetEventResponse> {
        return post(
            fn = "getEvent",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = GetEventRequest.serializer(),
            respSerializer = GetEventResponse.serializer()
        )
    }

    fun createEvent(
        idToken: String,
        adminToken: String?,
        payload: CreateEventRequest
    ): CallResult<CreateEventResponse> {
        return post(
            fn = "createEvent",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CreateEventRequest.serializer(),
            respSerializer = CreateEventResponse.serializer()
        )
    }

    fun publishEvent(
        idToken: String,
        adminToken: String?,
        payload: PublishEventRequest
    ): CallResult<PublishEventResponse> {
        return post(
            fn = "publishEvent",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = PublishEventRequest.serializer(),
            respSerializer = PublishEventResponse.serializer()
        )
    }

    fun signupForEvent(
        idToken: String,
        adminToken: String?,
        payload: SignupForEventRequest
    ): CallResult<SignupForEventResponse> {
        return post(
            fn = "signupForEvent",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = SignupForEventRequest.serializer(),
            respSerializer = SignupForEventResponse.serializer()
        )
    }

    fun cancelEventSignup(
        idToken: String,
        adminToken: String?,
        payload: CancelEventSignupRequest
    ): CallResult<CancelEventSignupResponse> {
        return post(
            fn = "cancelEventSignup",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CancelEventSignupRequest.serializer(),
            respSerializer = CancelEventSignupResponse.serializer()
        )
    }

    fun claimEventOffer(
        idToken: String,
        adminToken: String?,
        payload: ClaimEventOfferRequest
    ): CallResult<ClaimEventOfferResponse> {
        return post(
            fn = "claimEventOffer",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = ClaimEventOfferRequest.serializer(),
            respSerializer = ClaimEventOfferResponse.serializer()
        )
    }

    fun checkInEvent(
        idToken: String,
        adminToken: String?,
        payload: CheckInEventRequest
    ): CallResult<CheckInEventResponse> {
        return post(
            fn = "checkInEvent",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CheckInEventRequest.serializer(),
            respSerializer = CheckInEventResponse.serializer()
        )
    }

    fun createEventCheckoutSession(
        idToken: String,
        adminToken: String?,
        payload: CreateEventCheckoutSessionRequest
    ): CallResult<CreateEventCheckoutSessionResponse> {
        return post(
            fn = "createEventCheckoutSession",
            idToken = idToken,
            adminToken = adminToken,
            payload = payload,
            reqSerializer = CreateEventCheckoutSessionRequest.serializer(),
            respSerializer = CreateEventCheckoutSessionResponse.serializer()
        )
    }
}
